frame/benchmarking/src/weights.rs

```diff
@@ -57,76 +59,96 @@ pub trait WeightInfo {
 	}
+	/// The range of component `i` is `[1, 100]`.
 	fn sr25519_verification(i: u32, ) -> Weight {
-		Weight::from_ref_time(319_000 as u64)
-			// Standard Error: 8_000
-			.saturating_add(Weight::from_ref_time(47_171_000 as u64).saturating_mul(i as u64))
+		Weight::from_ref_time(48_224_000 as u64)
+			// Standard Error: 18_799
+			.saturating_add(Weight::from_ref_time(46_506_091 as u64).saturating_mul(i as u64))
 	}
 	// Storage: Skipped Metadata (r:0 w:0)
+	/// The range of component `i` is `[0, 1000]`.
 	fn storage_read(i: u32, ) -> Weight {
-		Weight::from_ref_time(0 as u64)
-			// Standard Error: 3_000
-			.saturating_add(Weight::from_ref_time(2_110_000 as u64).saturating_mul(i as u64))
+		Weight::from_ref_time(127_000 as u64)
+			// Standard Error: 3_908
+			.saturating_add(Weight::from_ref_time(1_933_926 as u64).saturating_mul(i as u64))
 			.saturating_add(T::DbWeight::get().reads((1 as u64).saturating_mul(i as u64)))
 	}
 	// Storage: Skipped Metadata (r:0 w:0)
+	/// The range of component `i` is `[0, 1000]`.
 	fn storage_write(i: u32, ) -> Weight {
-		Weight::from_ref_time(0 as u64)
-			// Standard Error: 0
-			.saturating_add(Weight::from_ref_time(372_000 as u64).saturating_mul(i as u64))
+		Weight::from_ref_time(141_000 as u64)
+			// Standard Error: 330
+			.saturating_add(Weight::from_ref_time(377_734 as u64).saturating_mul(i as u64))
 			.saturating_add(T::DbWeight::get().writes((1 as u64).saturating_mul(i as u64)))
 	}
 }
 
 // For backwards compatibility and tests
 impl WeightInfo for () {
+	/// The range of component `i` is `[0, 1000000]`.
 	fn addition(_i: u32, ) -> Weight {
-		Weight::from_ref_time(103_000 as u64)
+		Weight::from_ref_time(119_000 as u64)
 	}
+	/// The range of component `i` is `[0, 1000000]`.
 	fn subtraction(_i: u32, ) -> Weight {
-		Weight::from_ref_time(105_000 as u64)
+		Weight::from_ref_time(112_000 as u64)
 	}
+	/// The range of component `i` is `[0, 1000000]`.
 	fn multiplication(_i: u32, ) -> Weight {
-		Weight::from_ref_time(113_000 as u64)
+		Weight::from_ref_time(118_000 as u64)
 	}
+	/// The range of component `i` is `[0, 1000000]`.
 	fn division(_i: u32, ) -> Weight {
-		Weight::from_ref_time(102_000 as u64)
+		Weight::from_ref_time(122_000 as u64)
 	}
-	fn hashing(_i: u32, ) -> Weight {
-		Weight::from_ref_time(20_865_902_000 as u64)
+	/// The range of component `i` is `[0, 100]`.
+	fn hashing(i: u32, ) -> Weight {
+		Weight::from_ref_time(21_342_134_000 as u64)
+			// Standard Error: 39_648
+			.saturating_add(Weight::from_ref_time(823_058 as u64).saturating_mul(i as u64))
 	}
+	/// The range of component `i` is `[1, 100]`.
 	fn sr25519_verification(i: u32, ) -> Weight {
-		Weight::from_ref_time(319_000 as u64)
-			// Standard Error: 8_000
-			.saturating_add(Weight::from_ref_time(47_171_000 as u64).saturating_mul(i as u64))
+		Weight::from_ref_time(48_224_000 as u64)
+			// Standard Error: 18_799
+			.saturating_add(Weight::from_ref_time(46_506_091 as u64).saturating_mul(i as u64))
 	}
```

Comment:
Okay, so I think I found an issue @koute @ggwpez 

If you see these results, the base weight is now `48_224_000`, and the per `i` weight is `46_506_091`. This is because each `sr25519` verification is around `4X_000_000` weight.

Based on the changes here: https://github.com/paritytech/substrate/issues/11804

We now take the minimum weight among all benchmarks, and make that the y intercept.

In this case, the benchmark runs between `[1, 100]`, which means there was never a case where we tested this benchmark at `0`, and thus, the minimum weight is when `i = 1`, and the slope is also this, so we are basically double counting. In this case, the y intercept SHOULD have been near 0, since this function does nothing, but we never tested for that.

Potential Solutions:

- Some way to cleverly detect this kind of situation, and handle it in the output results, solve the problem for everyone.
- We can change the range to `[0, 100]` rather than `[1, 100]`... but there are probably many situations where users do not start their benchmarks at a strictly 0 value.
- Have each benchmark run explicitly where each component is zero, to detect a proper minimum execution time... but not all benchmarks may execute successfully without the designer having that in mind.

Reply:
Hmmm.... this is a fair point! Although I don't think we can do much to handle this case automatically without making *some* sort of a tradeoff which is going to be wrong in certain situations.

There isn't really a way for the benchmarking machinery to know how the benchmark will behave at `0` without actually running it at `0`. Sure, it can linearly extrapolate that information, and in this case it would be right! But in other cases it's going to be wrong, depending on whether the data is actually perfectly linear, on the range of input and output values, etc. So to me this is more of a question of which tradeoff do we want to pick. The consequence of the weights being too high is a lower throughput of the chain (which *is* bad); the consequence of the weights being too low is a potential for a denial of service attack and/or accidental overload of the chain (which I'd argue is worse).

Personally I'd probably just go for changing the range of the benchmark to `[0, 100]`. I don't think it's a huge thing to ask the users to run their benchmarks over the *whole* input domain of their inputs? Especially the minimum and the maximum boundary. If I may exaggerate a little bit and take this to the extreme, supporting the use case of "I want to have this function work in the range of [0, 10000000] while I only benchmark it/test it in the range of [100, 200]" would be quite weird, wouldn't it? So I think it's pretty natural to just require the whole input domain to be covered by the benchmarks.

@ggwpez Your thoughts?

Reply:
> There isn't really a way for the benchmarking machinery to know how the benchmark will behave at 0 without actually running it at 0.  

Exactly. The developer of the benchmark has to ensure that he always puts the legal min values in the component range.    
We cannot hack around such a shortcoming in FRAME.  
Otherwise the benchmarking will assume that `1` is the lowest possible value and therefore use it to calculate the minimum.

Reply:
Yes I think so. @shawntabrizi any objections?  

> I can go through our benchmarks and fix them

If you want to - yes sure! Otherwise we can create an issue for it.

Reply:
this makes me question the new algorithm we are using, presuming the slowest extrinsic is the base extrinsic weight. I wonder if there are ways we can correct for negative numbers without introducing such a large assumption...

But yeah, i mean also correcting the ranges seems okay for now.

Reply:
> I wonder if there are ways we can correct for negative numbers without introducing such a large assumption...

Well, one hack we could do is to only do this if the base weight is negative, but this seems like only a partial fix to me. The linear regression *can* still generate a base weight of zero, or a base weight of `1`, or `10`, or `100`, or another low number which is lower than the *actual* base weight if the extrinsic were to be executed with its components set to `0`.

I definitely agree with you that this is not perfect, and it would be nice to do *something* better. But I don't really see what that would be without introducing other serious limitations.

Considering the main problem here is that the users might not run their benchmarks at `0` to get a true base weight maybe we could by default emit an error like "the components' range do not cover `0`; are you sure this is what you want?" and have a longer explanation of the problem, and allow the users to disable the check with an `#[attribute]` of some sort? Then people would either correct their benchmarks, or opt-in that the minimums are actual true minimums. (I'm not entirely sure if this is a good idea though; we'd probably have to empirically test for how many benchmarks per your average chain this would trigger, and how many benchmarks would have to be corrected.)

> But yeah, i mean also correcting the ranges seems okay for now.

Okay, great! I'll whip up a PR extending the ranges then.

Reply:
> actual base weight if the extrinsic were to be executed with its components set to 0.

All extrinsics have a base weight added to them, so it is possible that an extrinsic with components 0 would give a weight of 0.